ui/ts/components/table.ts

```diff
@@ -41,41 +58,101 @@ module Components {
        * The columns in the table will be displayed in the same order as the
        * returned array.
        */
-      columns: () => TableColumn<T>[];
+      columns: Utils.ReadOnlyProperty<TableColumn<T>[]>;
       /* rows is a function that returns an array of row data for display in the
        * table. 
-       *
-       * Rows will be displayed in the same order as the returned array.
        */
-      rows: () => T[];
+      rows: Utils.ReadOnlyProperty<T[]>;
     }
 
     class Controller<T> {
       data: TableData<T>;
+      sortedRows: Utils.ReadOnlyProperty<T[]>;
+      private _sortColumn: Utils.Property<TableColumn<T>> = Utils.Prop(null);
+      private _sortAscend: Utils.Property<boolean> = Utils.Prop(false);
 
       constructor(data: TableData<T>) {
         this.data = data;
+
+        this.sortedRows = Utils.Computed(data.rows, this._sortColumn, this._sortAscend, (rows: T[], sortCol: TableColumn<T>, asc: boolean): T[] => {
+          let result = _(rows);
+          if (sortCol && sortCol.sortable) {
+            // Sort the rows using the currently selected column, if the column
+            // is sortable. Sort by the output of the column's sortValue()
+            // method if specified, using the output of view() otherwise.
+            if (sortCol.sortValue) {
+              result = result.sortBy(sortCol.sortValue);
+            } else {
+              result = result.sortBy(sortCol.view);
+            }
+
+            if (asc) {
+              result = result.reverse();
+            }
+          }
+          return result.value();
+        });
+      };
+
+      /**
+       * SetSortColumn sets the column which is currently used for sorting
+       * purposes. 
+       *
+       * When setting a new sort column, the sort direction is always ascending.
+       * If the same column is set again, the sort direction is reversed to
+       * ascending.
+       */
+      SetSortColumn(col: TableColumn<T>): void {
+        if (!col.sortable) {
+          return;
+        }
+        if (this._sortColumn() !== col) {
+          this._sortColumn(col);
+          this._sortAscend(false);
+        } else {
+          this._sortAscend(!this._sortAscend());
+        }
+      }
+
+      /**
+       * IsSortColumn returns true if the provided column is the column
+       * currently used for sorting.
+       */
+      IsSortColumn(col: TableColumn<T>): boolean {
+        return this._sortColumn() === col;
       }
 
       /**
-       * renderHeaders returns a mithril element which contains the header row
+       * RenderHeaders returns a mithril element which contains the header row
        * for the table.
        */
-      renderHeaders(): MithrilElement {
+      RenderHeaders(): MithrilElement {
         let cols = this.data.columns();
-        let renderedCols = cols.map((col: TableColumn<T>) => m("th", col.title));
+        let sortClass = "sorted" + (this._sortAscend() ? " ascending" : "");
```

Comment:
Descending is the default because that's the default direction of javascript's sort functions. Ascending means a reversal is applied.  It could be called "reversed", I suppose.